### ISSUE TITLE
Pi 5 with built-in RTC throws error ioctl(4, RTC_UIE_ON, 0): Invalid argument

### ISSUE THREAD

**Issue by by** [raspberrypi/linux#5793]
### Describe the bug

hwclock -v throws msg. with error
Waiting for clock tick...
ioctl(4, RTC_UIE_ON, 0): Invalid argument
Waiting in loop for time from /dev/rtc0 to change
...got clock tick
on Pi 5 like on the previous Pis despite the fact that it now has a built-in RTC, so no RTC extension board from other manufacturers.

### Steps to reproduce the behaviour

Run hwclock -v on Pi 5

### Device (s)

Other

### System

cat /etc/rpi-issue
Raspberry Pi reference 2023-10-10
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 962bf483c8f326405794827cce8c0313fd5880a8, stage4

vcgencmd version
2023/12/14 16:43:25 
Copyright (c) 2012 Broadcom
version 0f881224 (release) (embedded)

uname -a
Linux aion 6.6.5-v8-16k+ #1706 SMP PREEMPT Fri Dec  8 18:45:36 GMT 2023 aarch64 GNU/Linux

**Comment by popcornmix** [comment]
Based on discussion in https://github.com/raspberrypi/linux/issues/3282 perhaps setting `CONFIG_RTC_INTF_DEV_UIE_EMUL` would avoid this?

Testing... yes - it avoids the ioctl complaint:
```
pi@pi5:~ $ sudo hwclock -v
hwclock from util-linux 2.38.1
System Time: 1702640345.932417
Trying to open: /dev/rtc0
Using the rtc interface to the clock.
Assuming hardware clock is kept in UTC time.
Waiting for clock tick...
...got clock tick
Time read from Hardware Clock: 2023/12/15 11:39:06
Hw clock time : 2023/12/15 11:39:06 = 1702640346 seconds since 1969
Time since last adjustment is 1702640346 seconds
Calculated Hardware Clock drift is 0.000000 seconds
2023-12-15 11:39:05.921739+00:00
```

**Comment by popcornmix** [comment]
@pelwell any thoughts? The docs do say:

> The emulation code will read the time from the hardware clock several times per second, please enable this option only if you know that you really need it.

**Comment by pelwell** [comment]
Aside from the ugly error, does it work without UIE support?

**Comment by popcornmix** [comment]
It works for the obvious use cases (reading the time and waking up Pi at an specified time).

@by are you missing any functionality due to this?

**Comment by by** [comment]
No, haven't found functional errors yet; as with the Pi 5, RTC became an official product feature, it would be nice to correct the ugly error.

And yes, the kernel switch avoids it, and it would be nice to get it set in the 6.x kernels, similarly to https://github.com/raspberrypi/linux/pull/5057 , if the root cause can't be eliminated, thanks.

**Comment by rfrankla** [comment]
> Based on discussion in https://github.com/raspberrypi/linux/issues/3282 perhaps setting CONFIG_RTC_INTF_DEV_UIE_EMUL would avoid this?

Hi, err... that may not be a good idea. Below is from a CM4 on a Waveshare carrier board, with `Linux time-1 6.12.75+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.12.75-1+rpt1 (2026-03-11) aarch64`

```
rick@time-1:~ $ sudo strace hwclock -r 2> >(grep "^ioctl(4, RTC_RD_TIME" | wc -l)
2026-03-24 18:46:06.096735+00:00
615 
```
That's 615 `ioctl` calls just to find the edge of the next second. The exact number depends on when the call is made within a one second interval. If this were used in an NTP server (chronyd, ntpd, ...) as a 'local' rtc source w/'rtcfile' this may put a high load on the system, and increase the amount of non-determistic activities. These are negative attributes, in considering this proposal.

The 1Hz interrupt function is really needed for NTP servers using 'rtcfile', but they can survive using 'rtcsync'. But, if you want to use highly accurate, separate RTC chips like the RV-3028 as true local time sources (reference clock), then you really need real interrupts working.

BUT, the good news. I believe the root cause -- *missing functionality* -- in the rtc drivers may be fixed. Referring to the Datasheets for existing drivers...
Like used in: Uputronics GPS/RTC Expansion Board for Raspberry Pi 
https://www.microcrystal.com/fileadmin/Media/Products/RTC/App.Manual/RV-3028-C7_App-Manual.pdf

> 4.4.1.CLKOUT FREQUENCY SELECTION
A programmable square wave is available at pin CLKOUT. Operation is controlled by the FD field (EEPROM 35h).
Frequencies from 32.768 kHz (Default value on delivery) to 1 Hz and countdown timer interrupt can be generated for
use as a system clock, microcontroller clock, input to a charge pump, or for calibration of the crystal oscillator.

Dallas Semiconductor DS1307 Serial Real-Time Clock
> SQW/OUT (Square Wave/Output Driver) – When enabled, the SQWE bit set to 1, the SQW/OUT pin outputs one of four square wave frequencies (**1Hz**, 4kHz, 8kHz, 32kHz). 

and for RPI 5 the datasheet for DA9091   ...is not released!    (although it is likely to be similar to earlier Dialog/Renesas products)
https://www.renesas.com/en/document/dst/da9062-datasheet

> **8.13 Real-Time Clock** The alarm can alternatively be asserted from a periodic tick signal that, depending on register TICK_TYPE, is either asserted every second or minute. 

I will first work to fix the RV3028 driver, because I have everything here to do the job.
I will then fix the DS1307 driver (or another cheap failing RTC chip) to show it can be done.
I will then buy an RPI 5, and try to fix the DA9091 driver, with or without the datasheet.

Thank you for reading. Rick.
